File: bindings/python/aster/peer_store.py
"""
aster.peer_store -- Per-peer admission attribute store.

Bridges the gap between admission (where attributes are determined)
and RPC dispatch (where attributes are needed for authorization).

Both consumer admission and delegated admission handlers write to this
store on successful admission. The RPC server reads from it when
building CallContext for each call.

Expiry: each admission records the credential's ``expires_at`` (epoch
seconds). Lookups check expiry and lazily evict stale entries. A
background reaper sweeps entries that are never looked up again.

Config: ``ASTER_PEER_TTL_S`` (env var) sets the server-side upper bound
on how long a peer stays admitted, regardless of credential expiry.
Default 86400 (24 hours).
"""

from __future__ import annotations

import asyncio
import logging
import os
import time
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)

PEER_TTL_S: float = float(os.environ.get("ASTER_PEER_TTL_S", "86400"))
"""Server-side upper bound on peer admission lifetime (seconds).
Applied regardless of credential expiry. Whichever is sooner wins."""

_REAPER_INTERVAL_S: float = 300.0


@dataclass
class PeerAdmission:
    """Record of a successful peer admission."""

    endpoint_id: str
    handle: str = ""
    attributes: dict[str, str] = field(default_factory=dict)
    admitted_at: float = field(default_factory=time.time)
    expires_at: float = 0.0
    admission_path: str = ""  # "consumer_admission" | "aster.admission"

    def is_expired(self) -> bool:
        now = time.time()
        if self.expires_at > 0 and now > self.expires_at:
            return True
        if now > self.admitted_at + PEER_TTL_S:
            return True
        return False


class PeerAttributeStore:
    """In-memory store mapping peer endpoint_id to admission attributes.

    Entries are lazily evicted on access when expired. A background
    reaper task (started via ``start_reaper()``) sweeps entries that
    are never accessed again.
    """

    def __init__(self) -> None:
        self._peers: dict[str, PeerAdmission] = {}
        self._reaper_task: asyncio.Task | None = None

    def admit(self, admission: PeerAdmission) -> None:
        """Record a successful admission."""
        self._peers[admission.endpoint_id] = admission

    def get(self, endpoint_id: str) -> PeerAdmission | None:
        """Look up admission record for a peer. Returns None if expired."""
        admission = self._peers.get(endpoint_id)
        if admission is None:
            return None
        if admission.is_expired():
            self._peers.pop(endpoint_id, None)
            logger.debug("Peer %s expired (admitted_at=%.0f, expires_at=%.0f)",
                         endpoint_id, admission.admitted_at, admission.expires_at)
            return None
        return admission

    def remove(self, endpoint_id: str) -> None:
        """Remove a peer on disconnect or revocation."""
        self._peers.pop(endpoint_id, None)

    def get_attributes(self, endpoint_id: str) -> dict[str, str]:
        """Get attributes dict for a peer, or empty if expired/not admitted."""
        admission = self.get(endpoint_id)
        return dict(admission.attributes) if admission else {}

    def sweep_expired(self) -> int:
        """Remove all expired entries. Returns count removed."""
        expired = [eid for eid, adm in self._peers.items() if adm.is_expired()]
        for eid in expired:
            self._peers.pop(eid, None)
        if expired:
            logger.debug("Reaped %d expired peer admissions", len(expired))
        return len(expired)

    @property
    def peer_count(self) -> int:
        return len(self._peers)

    def start_reaper(self) -> None:
        """Start background reaper task (call once at server start)."""
        if self._reaper_task is not None and not self._reaper_task.done():
            return
        self._reaper_task = asyncio.get_event_loop().create_task(self._reaper_loop())

    def stop_reaper(self) -> None:
        if self._reaper_task is not None:
            self._reaper_task.cancel()
            self._reaper_task = None

    async def _reaper_loop(self) -> None:
        try:
            while True:
                await asyncio.sleep(_REAPER_INTERVAL_S)
                self.sweep_expired()
        except asyncio.CancelledError:
            pass

File: bindings/typescript/packages/aster/src/peer-store.ts
/**
 * aster/peer-store -- Per-peer admission attribute store.
 *
 * Bridges the gap between admission (where attributes are determined)
 * and RPC dispatch (where attributes are needed for authorization).
 *
 * Expiry: each admission records the credential's `expiresAt` (epoch
 * seconds). Lookups check expiry and lazily evict stale entries. A
 * background reaper sweeps entries that are never looked up again.
 *
 * Config: ASTER_PEER_TTL_S env var (default 86400 = 24h) sets the
 * server-side upper bound on how long a peer stays admitted.
 */

/** Server-side upper bound on peer admission lifetime (seconds). */
export const PEER_TTL_S: number = (() => {
  if (typeof process !== 'undefined' && process.env?.ASTER_PEER_TTL_S) {
    return parseFloat(process.env.ASTER_PEER_TTL_S);
  }
  return 86400;
})();

const _REAPER_INTERVAL_MS = 300_000;

/** Record of a successful peer admission. */
export interface PeerAdmission {
  endpointId: string;
  handle: string;
  attributes: Map<string, string>;
  admittedAt: number;
  expiresAt: number;
  /** "consumer_admission" | "aster.admission" */
  admissionPath: string;
}

/** Create a PeerAdmission with defaults for optional fields. */
export function createPeerAdmission(
  opts: Pick<PeerAdmission, 'endpointId'> & Partial<Omit<PeerAdmission, 'endpointId'>>,
): PeerAdmission {
  return {
    endpointId: opts.endpointId,
    handle: opts.handle ?? '',
    attributes: opts.attributes ?? new Map(),
    admittedAt: opts.admittedAt ?? Date.now() / 1000,
    expiresAt: opts.expiresAt ?? 0,
    admissionPath: opts.admissionPath ?? '',
  };
}

function isExpired(admission: PeerAdmission): boolean {
  const now = Date.now() / 1000;
  if (admission.expiresAt > 0 && now > admission.expiresAt) return true;
  if (now > admission.admittedAt + PEER_TTL_S) return true;
  return false;
}

/**
 * In-memory store mapping peer endpointId to admission attributes.
 *
 * Entries are lazily evicted on access when expired. Call startReaper()
 * to enable periodic sweeps of entries that are never accessed.
 */
export class PeerAttributeStore {
  private peers: Map<string, PeerAdmission> = new Map();
  private reaperTimer: ReturnType<typeof setInterval> | null = null;

  /** Record a successful admission. */
  admit(admission: PeerAdmission): void {
    this.peers.set(admission.endpointId, admission);
  }

  /** Look up admission record for a peer. Returns undefined if expired. */
  get(endpointId: string): PeerAdmission | undefined {
    const admission = this.peers.get(endpointId);
    if (!admission) return undefined;
    if (isExpired(admission)) {
      this.peers.delete(endpointId);
      return undefined;
    }
    return admission;
  }

  /** Remove a peer on disconnect or revocation. */
  remove(endpointId: string): void {
    this.peers.delete(endpointId);
  }

  /** Get attributes map for a peer, or empty map if expired/not admitted. */
  getAttributes(endpointId: string): Map<string, string> {
    const admission = this.get(endpointId);
    return admission ? new Map(admission.attributes) : new Map();
  }

  /** Remove all expired entries. Returns count removed. */
  sweepExpired(): number {
    let count = 0;
    for (const [eid, adm] of this.peers) {
      if (isExpired(adm)) {
        this.peers.delete(eid);
        count++;
      }
    }
    return count;
  }

  get peerCount(): number {
    return this.peers.size;
  }

  /** Start background reaper (call once at server start). */
  startReaper(): void {
    if (this.reaperTimer != null) return;
    this.reaperTimer = setInterval(() => this.sweepExpired(), _REAPER_INTERVAL_MS);
    if (typeof this.reaperTimer === 'object' && 'unref' in this.reaperTimer) {
      (this.reaperTimer as any).unref();
    }
  }

  /** Stop background reaper. */
  stopReaper(): void {
    if (this.reaperTimer != null) {
      clearInterval(this.reaperTimer);
      this.reaperTimer = null;
    }
  }
}
